extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts

```diff
@@ -0,0 +1,398 @@
+        }
+    }
+
+    // @installed
+    // @builtin
+    // @enabled
+    // @disabled
+    hasType(filter: string, type: string) {
+        if (filter) {
+            const filters = filter.split(' ');
+            const found = filters.find(value => value === type);
+            return found !== undefined;
+        }
+
+        return false;
+    }
+
+    filterByType(plugins: ChePluginMetadata[], type: string): ChePluginMetadata[] {
+        return plugins.filter(plugin => {
+            const regex = / /gi;
+            const t = plugin.type.toLowerCase().replace(regex, '_');
+            return t === type;
+        });
+    }
+
+    filterByText(plugins: ChePluginMetadata[], text: string): ChePluginMetadata[] {
+        return plugins;
+    }
+
+    filter(plugins: ChePluginMetadata[], filter: string): ChePluginMetadata[] {
+        let filteredPlugins = plugins;
+        const filters = filter.split(' ');
+
+        filters.forEach(f => {
+            if (f) {
+                if (f.startsWith('@')) {
+                    if (f.startsWith('@type:')) {
+                        const type = f.substring('@type:'.length);
+                        filteredPlugins = this.filterByType(filteredPlugins, type);
+                    }
+                } else {
+                    filteredPlugins = this.filterByText(filteredPlugins, f);
+                }
+            }
+        });
+
+        return filteredPlugins;
+    }
+
+    /**
+     * Removes plugins with type 'Che Editor' if type '@type:che_editor' is not set
+     */
+    squeezeOutEditors(plugins: ChePluginMetadata[], filter: string): ChePluginMetadata[] {
+        // do not filter if user requested list of editors
+        if (this.hasType(filter, '@type:che_editor')) {
+            return plugins;
+        }
+
+        return plugins.filter(plugin => 'Che Editor' !== plugin.type);
+    }
+
+    /**
+     * Returns a list of available plugins on the plugin registry.
+     *
+     * @param registry ChePluginRegistry plugin registry
+     * @param filter filter
+     * @return list of available plugins
+     */
+    async getPlugins(registry: ChePluginRegistry, filter: string): Promise<ChePluginMetadata[]> {
+        // ensure default plugin registry URI is set
+        if (!this.defaultRegistry) {
+            await this.getDefaultRegistry();
+        }
+
+        let pluginList;
+        if (filter) {
+            if (this.hasType(filter, '@installed')) {
+                pluginList = await this.getInstalledPlugins();
+            } else {
+                pluginList = await this.getAllPlugins(registry);
+            }
+
+            pluginList = this.filter(pluginList, filter);
+        } else {
+            pluginList = await this.getAllPlugins(registry);
+        }
+
+        // remove che_editor if type @type:che_editor is not set
+        pluginList = this.squeezeOutEditors(pluginList, filter);
+
+        return pluginList;
+    }
+
+    // without prefix
```

Comment:
Could you please write a bit more clear explanation

Reply:
Done